### PR TITLE
Migrate DataPayload to CartableOptionPointer

### PR DESCRIPTION
**Pull request by sffc**
Depends on #4449
Depends on #4456
Replaces #4442
Part of #4438

### CODE REVIEW

**Comment by sffc** [comment]
The first, and primary, commit on this branch is 0eb4966. The other commits are inherited from the dependent PRs.

File: provider/core/src/response.rs [hunk 1]
```diff
@@ -106,6 +107,17 @@ impl Cart {
             .map(|yoke| unsafe { yoke.replace_cart(Cart) })
             .map(Yoke::wrap_cart_in_option)
     }
+
+    #[inline]
+    pub(crate) fn unwrap_yoke<Y>(
```
Comment (Manishearth) [inline provider/core/src/response.rs]: question: do we still need this `Cart` type?

File: provider/core/src/response.rs [hunk 2]
```diff
@@ -106,6 +107,17 @@ impl Cart {
...
+        yoke: Yoke<Y, Option<Cart>>,
+    ) -> Yoke<Y, Option<SelectedRc<Box<[u8]>>>>
+    where
+        for<'a> Y: Yokeable<'a>,
+    {
+        // Safety: `Cart` has one field and we are removing it from the newtype.
```
Comment (Manishearth) [inline provider/core/src/response.rs:119]: "and we are preserving it in the new cart, unwrapping it from the newtype"

File: provider/core/src/response.rs (hunk 1 above)
Comment (Manishearth) [reply provider/core/src/response.rs]: ah we have a pub type here. I see.

**Review by Manishearth — COMMENTED** [review]
LGTM overall, would probably wait for dependents to filter through to properly approve

File: provider/core/src/response.rs (hunk 2 above)
Comment (Manishearth) [reply provider/core/src/response.rs:119]: Also I'd call this function `unwrap_cart` or something, and add docs.

**Review by Manishearth — COMMENTED** [review]
lgtm except for changes to unwrap_yoke

**Review by Manishearth — APPROVED** [review]

**Pull request merged**